core/src/main/java/com/datastax/oss/driver/internal/core/context/LifecycleListener.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.context;
+
+import com.datastax.oss.driver.api.core.session.SessionBuilder;
+
+/**
+ * A component that gets notified of certain events in the session's lifecycle.
+ *
+ * <p>This is intended for third-party extensions, no built-in components implement this.
+ */
+public interface LifecycleListener extends AutoCloseable {
+
+  /**
+   * Invoked when the session is ready to process user requests.
+   *
+   * <p>This corresponds to the moment when the {@link SessionBuilder#build()} returns, or the
+   * future returned by {@link SessionBuilder#buildAsync()} completes. If the session initialization
+   * fails, this method will not get called.
+   *
+   * <p>This method is invoked on a driver thread, it should complete relatively quickly and not
+   * block.
+   */
+  void onSessionReady();
+
+  /**
+   * Invoked when the session shuts down.
+   *
+   * <p>Note that this method gets called even if the shutdown results from a failed initialization.
+   * In that case, implementations should be ready to handle a call to this method even though
+   * {@link #onSessionReady()} hasn't been invoked.
+   *
+   * <p>This method is invoked on a driver thread, it should complete relatively quickly and not
+   * block.
+   */
+  @Override
+  void close() throws Exception;
```

Comment:
Is this meant to be where implementors shut down their own listeners, or as a notification hook that gets invoked when the session closes? For clarity I think we should dissociate both and add a method `onSessionClosed` to capture the second use case. Also, we might want to pass the `Throwable` that caused the session initialization to fail.

Reply:
It gets called when the session closes.
Implementors should perform any necessary cleanup. A typical example would be to cancel a scheduled task that they had scheduled on a driver admin event loop.

Reply:
I don't understand how `onSessionClosed` would be a different use case, both methods would be called when the session closes.
As for passing the `Throwable` that caused an init failure, I don't see a strong case for it. This method mostly exists so that we don't leave anything running on the admin event loops, I think it's unlikely that a listener would do something differently depending on the error.
And if we find the need to add it later, it will be relatively easy to have a default implementation of `close(Throwable)` that calls `close()`.

Reply:
It's a matter os semantics: `onSessionClosed` means: "I am being notified that a given session has been closed" whereas the contract of `close` (as defined in `AutoCloseable`) is "I am being told to close myself". Here, it sounds to me as if we are distorting the original purpose of `close`.

More generally speaking, a typical listener API would expose methods such as `onEventXSuccessful(Foo receiver)`, `onEventYFailed(Bar receiver, Throwable error)`. It would also expose methods to register/unregister the listener against a given Session. None of this is present in this interface, which is why I was a bit thrown off initially.

But OK to keep things simple for now; it's an internal API after all so we are free to change it later.